### PR TITLE
Add GA4 client ID to stub attribution

### LINKED ISSUE #12801: [GA4] Ensure GA client_id and session_id is passed to stub attribution
## Description

Using Universal Analytics we pass the GA client ID ~and session ID~ to stub attribution to help measure Firefox installs. We need to check that this still works as expected in GA4, and if not work out what's possible.

The objects we currently rely on are:

- client id: `window.ga.getAll()[0].get('clientId')`
- ~session id: `window.ga.getAll()[0].get('_gt')`~

---

#### Success Criteria
- [ ] Understand if switching to GA4 impacts this measurement
- [ ] Identify possible remedy if needed.

### PR DESCRIPTION
**Pull request by stephaniehobson**
## One-line summary

Send GA4 client ID to stub attribution as `client_id_ga4`.

I'm sorry to say this has a bit of urgency. It cannot be merged until the stub attribution server is updated but we should get a jump start on reviewing it - especially since we want a VERY THROUGH review.

## Significant changes and points to review

- stub attribution now waits for both UA and GTM API before executing
- GA4 client ID is retrieved from GTM API
- data sent to stub attribution now includes the GA4 client ID as `client_id_ga4`
- updated and added tests
- tried to standardize on UA/GA4 when talking about the old and new versions
- tried to be explicit about when we're interacting with GTM rather than UA/GA4 directly
- changed references to scene2 to download/thanks
- added some steps for code review into the docs

## Issue / Bugzilla link

Fix #12801 

## Testing

I worked on this over a couple weeks so you should not assume I was on the top of my game while writing this code.

- currently deployed on demo9
- steps to test are included in the PR
- do we have good test coverage?

### CODE REVIEW

**Comment by stephaniehobson** [comment]
Some new tests are failing and I'm not sure why :(

File: media/js/base/stub-attribution.js [hunk 1]
```diff
@@ -369,28 +420,33 @@ if (typeof window.Mozilla === 'undefined') {
     };
 
     /**
-     * A crude check to see if Google Analytics has loaded. Periodically
-     * attempts to retrieve the client ID from the global `ga` object.
+     * A crude check to see if Google Analytics has loaded.
      * @param {Function} callback
      */
-    StubAttribution.waitForGoogleAnalytics = function (callback) {
```
Comment (alexgibson) [inline media/js/base/stub-attribution.js:376]: If you're renaming things then you might need to check this does not break the RTAMO code at https://github.com/mozilla/bedrock/blob/main/media/js/firefox/new/common/thanks-direct.js

File: media/js/base/stub-attribution.js [hunk 2]
```diff
@@ -434,8 +491,12 @@ if (typeof window.Mozilla === 'undefined') {
             ua: ua,
             experiment: experiment,
             variation: variation,
-            client_id: clientID,
-            session_id: clientID ? StubAttribution.createSessionID() : null,
+            client_id: clientIDUA,
+            client_id_ga4: clientIDGA4,
```
Comment (alexgibson) [inline media/js/base/stub-attribution.js:495]: If you're adding a new field to stub attribution then you also need to add it to the list of known fields in bedrock's back-end stub authentication service (this exists so that people can't encode any suspicious data) https://github.com/mozilla/bedrock/blob/main/bedrock/firefox/views.py#L46

This is probably why you don't see the new field when decoding what bedrock's back-end is returning.

File: tests/unit/spec/base/stub-attribution.js [hunk 3]
```diff
@@ -536,20 +554,66 @@ describe('stub-attribution.js', function () {
...
+                GA_CLIENT_ID
+            );
+            // stub out GA4 client ID
+            spyOn(Mozilla.StubAttribution, 'getGtagClientID').and.returnValue(
+                null
+            );
             const callback = jasmine.createSpy('callback');
 
-            Mozilla.StubAttribution.waitForGoogleAnalytics(callback);
+            Mozilla.StubAttribution.waitForGoogleAnalyticsThen(callback);
```
Comment (alexgibson) [inline tests/unit/spec/base/stub-attribution.js:584]: I think this code is failing because jasmine's own internal clock is not waiting for the code's `setTimeout()` to run its course. To fix this you can use [clock](https://jasmine.github.io/api/5.1/Clock.html) and then tick the clock forward, forcing the timeouts to execute. Something like:

```javascript
beforeEach(function () {
    jasmine.clock().install();
});

afterEach(function () {
    jasmine.clock().uninstall();
});

it('should fire a callback when UA client ID is found but GA4 client ID is not', async function () {
    // stub out UA client ID
    spyOn(Mozilla.StubAttribution, 'getUAClientID').and.returnValue(
        GA_CLIENT_ID
    );
    // stub out GA4 client ID
    spyOn(Mozilla.StubAttribution, 'getGtagClientID').and.returnValue(
        null
    );
    const callback = jasmine.createSpy('callback');

    const result = await Mozilla.StubAttribution.waitForGoogleAnalyticsThen(callback);
    jasmine.clock().tick(2100);
    expect(callback).toHaveBeenCalledWith(true);
});
```

The nice thing here is it doesn't actually slow down the tests :)

**Review by alexgibson — CHANGES_REQUESTED** [review]

File: tests/unit/spec/base/stub-attribution.js (hunk 3 above)
Comment (stephaniehobson) [reply tests/unit/spec/base/stub-attribution.js:584]: Fixed it, thanks!

File: media/js/base/stub-attribution.js (hunk 2 above)
Comment (stephaniehobson) [reply media/js/base/stub-attribution.js:495]: Working! Thanks!

File: media/js/base/stub-attribution.js [hunk 4]
```diff
@@ -357,6 +357,56 @@ if (typeof window.Mozilla === 'undefined') {
         }
     };
 
+    /**
+     * Attempts to retrieve the GA4 client and session ID from the dataLayer
```
Comment (willdurand) [inline media/js/base/stub-attribution.js:361]: We do not retrieve the session ID here, right?

Comment (stephaniehobson) [reply media/js/base/stub-attribution.js:361]: Correct. It's an outdated comment, I will update it. TY!

File: package-lock.json [hunk 5]
```diff
@@ -57,11 +57,11 @@
       }
     },
     "node_modules/@ampproject/remapping": {
-      "version": "2.2.0",
-      "resolved": "https://registry.npmjs.org/@ampproject/remapping/-/remapping-2.2.0.tgz",
-      "integrity": "sha512-qRmjj8nj9qmLTQXXmaR1cck3UXSRMPrbsLJAasZpF+t3riI71BXed5ebIOYwQntykeZuhjsdweEc9BxH5Jc26w==",
+      "version": "2.2.1",
```
Comment (alexgibson) [inline package-lock.json:60]: **issue:** these changes seem unrelated to this PR? If so, can we revert them?

**Review by alexgibson — CHANGES_REQUESTED** [review]
I tested things locally and everything seems to be working well, nice work! 

The Python tests are failing, so it looks like you'll need to update those still. Since we're adding a new param to all requests, the encoded values for `moz-stub-attribution-code` all change. So I think you'll need to just update the expected values to match.

When testing locally, I also noticed that my value for `client_id` is always identical to `client_id_ga4`? So it seems a little odd that we need to add a whole new field here to the service. My assumption here is that DS want this to be clean? (and that's totally fine). But attribution strings do have a fixed maximum length, so the more things we add it increases the chance of trimming campaign values and such.

I guess my only recommendation for that, is we should make sure we remove `client_id` once UA is turned off?

File: package-lock.json (hunk 5 above)
Comment (stephaniehobson) [reply package-lock.json:60]: Yes. WTF.

**Comment by stephaniehobson** [comment]
I don't know how to generate a new `attribution_sig`. Let's connect on Monday.

**Comment by alexgibson** [comment]
> I don't know how to generate a new attribution_sig. Let's connect on Monday.

I'm not sure how to generate the exact calls for each test manually, but the quickest way to update the expected results in each test failure. You can also base64 decode `attribution_code` to make sure the values look right as an extra measure if needed.

**Comment by stephaniehobson** [comment]
Tests are passing!
